content/blog/2025-03-20-new-hash-table.md

```diff
@@ -0,0 +1,202 @@
+                                   v
+                                  ...
+```
+
+The dict has two tables, called "table 0" and "table 1". Usually only one
+exists, but both are used when incremental rehashing is in progress.
+
+It is a chained hash table, so if multiple keys are hashed to the same slot in
+the table, their key-value entries form a linked list. That's what the "next"
+pointer in the dictEntry is for.
+
+To lookup a key "FOO" and access the value "BAR", Valkey still had to read from
+memory four times. If there is a hash collission, it has to follow two more
+pointers for each hash collission and thus read twice more from memory (the key
+and the next pointer).
+
+In Valkey 8.0, an optimization was made to embed the key ("FOO" in the drawing
+above) in the dictEntry, eliminating one pointer and one memory access.
+
+Design
+------
+
+In the new hash table designed for Valkey 8.1, the table consists of buckets of
+64 bytes, one cache line. Each bucket can store up to seven elements. Keys that
+map to the same bucket are all stored in the same bucket. The bucket also has a
+metadata section which contains a one byte secondary hash for each key. This is
+used for quickly eliminating hash collissions when looking up a key. In this
+way, we can avoid comparing the key for a mismatching key, except once in 256.
+
+We eliminated the dictEntry and instead embed key and value in the serverObject,
+along with other metadata for the key.
+
+```
++-----------+
+| hashtable |        bucket            bucket            bucket
++-----------+       +-----------------+-----------------+-----------------+-----
+| table 0  -------->| m x x x x x x x | m x x x x x x x | m x x x x x x x | ...
+| table 1   |       +-----------------+-----|-----------+-----------------+-----
++-----------+                               |
+                                            v
+                               +------------------------+
+                               | serverObject           |
+                               +------------------------+
+                               | type, encoding,        |
+                               | ref-counter, etc.      |
+                               | "FOO" (embedded key)   |
+                               | "BAR" (embedded value) |
+                               +------------------------+
+```
+
+Assuming the hashtable and the table are already in the CPU cache, looking up
+key-value entry now requires only two memory lookups: The bucket and the
+serverObject. If there is a hash collission, the object we're looking for is
+most likely in the same bucket, so no extra memory access is required.
+
+If a bucket becomes full, the last element slots in the bucket is replaced by a
+pointer to a child bucket. A child bucket has the same layout as a regular
+bucket, but it's a separate allocation. Child buckets form a chain. There is
+some probability of this happening, but long chains are very rare. Most of the
```

Comment:
```suggestion
bucket, but it's a separate allocation. The length of these bucket chains are not bounded, but long chains are very rare as long as keys are well distributed by the hashing function. Most of the
```

The, "there is some probability" just seems weird to me.

Reply:
Thanks. I'll take this with some line breaks. (I don't like long lines. :laughing:)

```suggestion
bucket, but it's a separate allocation. The length of these bucket chains are
not bounded, but long chains are very rare as long as keys are well distributed
by the hashing function. Most of the
```